kratos/processes/simple_mortar_mapper_process.h

```diff
@@ -0,0 +1,474 @@
+        {
+            auto it_cond = mrThisModelPart.ConditionsBegin() + i;
+            
+            if (it_cond->Is(SLAVE) == true)
+            {
+                const array_1d<double, 3>& slave_normal = it_cond->GetValue(NORMAL);
+                GeometryType& slave_geometry = it_cond->GetGeometry();
+                
+                boost::shared_ptr<ConditionMap>& all_conditions_maps = it_cond->GetValue( MAPPING_PAIRS );
+                
+                for (auto it_pair = all_conditions_maps->begin(); it_pair != all_conditions_maps->end(); ++it_pair )
+                {
+                    Condition::Pointer p_cond_master = (it_pair->first); // MASTER
+                    const array_1d<double, 3>& master_normal = p_cond_master->GetValue(NORMAL); 
+                    GeometryType& master_geometry = p_cond_master->GetGeometry();
+                        
+                    const double total_master_area = master_geometry.Area();
+                    double master_area = 0.0;
+                    
+                    IntegrationMethod this_integration_method = GeometryData::GI_GAUSS_2;
+                    
+                    // Reading integration points
+                    std::vector<array_1d<PointType,TDim>> conditions_points_slave;
+                    const bool is_inside = integration_utility.GetExactIntegration(slave_geometry, slave_normal, master_geometry, master_normal, conditions_points_slave);
+                    
+                    if (is_inside == true)
+                    {
+                        // Initialize general variables for the current master element
+                        rVariables.Initialize();
+                
+                        // Initialize the mortar operators
+                        rThisMortarConditionMatrices.Initialize();
+                        
+                        for (unsigned int i_geom = 0; i_geom < conditions_points_slave.size(); i_geom++)
+                        {
+                            std::vector<PointType::Pointer> points_array (TDim); // The points are stored as local coordinates, we calculate the global coordinates of this points
+                            for (unsigned int i_node = 0; i_node < TDim; i_node++)
+                            {
+                                PointType global_point;
+                                slave_geometry.GlobalCoordinates(global_point, conditions_points_slave[i_geom][i_node]);
+                                points_array[i_node] = boost::make_shared<PointType>(global_point);
+                            }
+                            
+                            typename std::conditional<TDim == 2, LineType, TriangleType >::type decomp_geom( points_array );
+                            
+                            const bool bad_shape = (TDim == 2) ? MortarUtilities::LengthCheck(decomp_geom, slave_geometry.Length() * 1.0e-6) : MortarUtilities::HeronCheck(decomp_geom);
```

Comment:
I am trying, but I really can not follow your algorithm. As I understand here you are creating the tessellatikn geometries. However it is really strange to me that you do it for every master touching a given slave and not the contrary.

It would have felt natural to me to do the splitting by looping over all the master conditions and for all of them on all the slave conditions touching the given master. That way the calculation would have been completely independent. I am not saying what you are doing is wrong, just that I can not follow the logics

Reply:
I am using the definition of mortar of Popp, so my slave is your master. What you say is basically the same, I skip some small operations, but it complicates further operations